transport/core/backoff.go

```diff
@@ -18,73 +38,161 @@ var DefaultInterval = 5 * time.Minute
+func New(max time.Duration, interval time.Duration) *Backoff {
+	if max < 0 || interval < 0 {
+		panic("backoff: max or interval is negative")
 	}
 
-	if b.MaxDuration == 0 {
-		b.MaxDuration = DefaultMaxDuration
+	b := &Backoff{
+		maxDuration: max,
+		interval:    interval,
 	}
+	b.setup()
+	return b
+}
 
-	if b.lock == nil {
-		b.lock = new(sync.Mutex)
+// NewWithoutJitter works similarly to New, except that the created
+// Backoff will not use jitter.
+func NewWithoutJitter(max time.Duration, interval time.Duration) *Backoff {
+	b := New(max, interval)
+	b.noJitter = true
+	return b
+}
+
+func init() {
+	var buf [8]byte
+	var n int64
+
+	_, err := io.ReadFull(rand.Reader, buf[:])
+	if err != nil {
+		panic(err.Error())
+	}
+
+	n = int64(binary.LittleEndian.Uint64(buf[:]))
+
+	src := mrand.NewSource(n)
+	prng = mrand.New(src)
+}
+
+func (b *Backoff) setup() {
+	if b.interval == 0 {
+		b.interval = DefaultInterval
+	}
+
+	if b.maxDuration == 0 {
+		b.maxDuration = DefaultMaxDuration
 	}
 }
 
 // Duration returns a time.Duration appropriate for the backoff,
 // incrementing the attempt counter.
 func (b *Backoff) Duration() time.Duration {
 	b.setup()
-	b.lock.Lock()
-	defer b.lock.Unlock()
 
-	pow := 1 << uint(b.tries)
+	b.decayN()
 
-	// MaxInt16 is an arbitrary choice on an upper bound; the
-	// implication is that every 16 tries, the counter resets.
-	if pow > math.MaxInt16 {
-		b.tries = 0
-		pow = 1
-	}
+	t := b.duration(b.n)
 
-	t := time.Duration(pow)
-	b.tries++
-	t = b.Interval * t
-	if t > b.MaxDuration {
-		t = b.MaxDuration
+	if b.n < math.MaxUint64 {
+		b.n++
 	}
 
-	if b.Jitter {
-		t = time.Duration(mrand.Int63n(int64(t)))
+	if !b.noJitter {
+		prngMu.Lock()
+		t = time.Duration(prng.Int63n(int64(t)))
+		prngMu.Unlock()
 	}
 
 	return t
 }
 
-// Reset clears the backoff.
+// requires b to be locked.
+func (b *Backoff) duration(n uint64) (t time.Duration) {
+	// Saturate pow
+	pow := time.Duration(math.MaxInt64)
+	if n < 63 {
+		pow = 1 << n
+	}
+
+	t = b.interval * pow
+	if t/pow != b.interval || t > b.maxDuration {
+		t = b.maxDuration
+	}
+
+	return
+}
+
+// Reset resets the attempt counter of a backoff.
+//
+// It should be called when the rate-limited action succeeds.
 func (b *Backoff) Reset() {
-	b.setup()
-	b.lock.Lock()
-	b.tries = 0
-	b.lock.Unlock()
+	b.lastTry = time.Time{}
+	b.n = 0
+}
+
+// SetDecay sets the duration after which the try counter will be reset.
+// Panics if decay is smaller than 0.
+//
+// The decay only kicks in if at least the last backoff + decay has elapsed
+// since the last try.
+func (b *Backoff) SetDecay(decay time.Duration) {
+	if decay < 0 {
+		panic("backoff: decay < 0")
+	}
+
+	b.decay = decay
+}
+
+// requires b to be locked
+func (b *Backoff) decayN() {
+	if b.decay == 0 {
+		return
+	}
+
+	if b.lastTry.IsZero() {
+		b.lastTry = time.Now()
+		return
+	}
+
+	lastDuration := b.duration(b.n - 1)
+	decayed := time.Since(b.lastTry) > lastDuration+b.decay
```

Comment:
This is discussed [in the paper](http://www.awsarchitectureblog.com/2015/03/backoff.html), if I recall correctly.